File: exercises/test_runner.py
debug_mode = False
def dprint(msg):
    if debug_mode:
        print(msg)
        
def run_test(solution_class_obj, solution_name, test_cases, debug: bool):

    import time
    
    debug_mode = debug
    solution_method = getattr(solution_class_obj, solution_name)

    for case_id in test_cases:
        case = test_cases[case_id]
        input = case['input']
        expected = case['expected']    
    
        start = time.perf_counter()
        result = solution_method(*input)
        success = result == expected
        end = time.perf_counter()
        exec_time = end-start
        print(f"{case_id} " + ("passed" if success else "failed") + f"\ntime taken: {exec_time}\n\n")
        if not success:
            print(f"Expected:\n\n{expected}\n\nResult:\n\n{result}\n\n")

        print("----------------------------\n\n")

# Setup
#

# from test_runner import run_test

# class ExampleExercise:
#     def sol1_desc(self, a):
#         pass


# test_cases = {
#     "case_1": {
#         "input": [...],
#         "expected": ...,
#     },
#     ...
# }

# run_test(
#     ExampleExercise(),
#     "sol1_desc",
#     test_cases,
#     False
# )

# Usage:
#
# Execute using the following:
#
# python3 -m <exercise_name>.<exercise_file>.py

File: exercises/test_runner.php
<?php

$debug_mode = false;

function dprint($msg) {
    global $debug_mode;
    if($debug_mode === true) {
        print($msg);
    }
}

function run_test(
    $solution_class_obj, 
    string $solution_name, 
    array $test_cases, 
    bool $debug,
    ?callable $resultOverride = null
) {
    global $debug_mode;
    $debug_mode = $debug;

    foreach($test_cases as $case_id => $case) {
        $input = $case['input'];
        $expected = $case['expected'];
    
        $start = microtime(true);
        $result = call_user_func_array([$solution_class_obj, $solution_name], $input);
        if($resultOverride) {
            $result = $resultOverride($result);
        }
        
        $success = $result === $expected;
        $end = microtime(true);
        $exec_time = $end-$start;
        print("{$case_id} " . ($success ? "passed" : "failed") . "\ntime taken: {$exec_time}\n\n");


        if(!$success) {
            print("Expected:\n\n".var_export($expected, true)."\n\nResult:\n\n".var_export($result, true)."\n\n");
        }

        print("----------------------------\n\n");
    }

}

// <?php

// require_once "test_runner.php";


// class Problem {

//     public function sol1_hashmap(...) {
//         return ...
//     }

// }

// $cases = [
//     "test1" => [
//         "input" => [
//             ..., 
//             ...
//         ],
//         "expected" => ...
//     ],
// ];


// run_test(new Problem(), "sol1", $cases, true);